piker/brokers/binance.py

```diff
@@ -0,0 +1,496 @@
+    https://binance-docs.github.io/apidocs/spot/en/#kline-candlestick-streams
+
+    """
+    time: int
+
+    open: float
+    high: float
+    low: float
+    close: float
+    volume: float
+
+    close_time: int
+
+    quote_vol: float
+    num_trades: int
+    buy_base_vol: float
+    buy_quote_vol: float
+    ignore: int
+
+    # null the place holder for `bar_wap` until we
+    # figure out what to extract for this.
+    bar_wap: float = 0.0
+
+
+# convert arrow timestamp to unixtime in miliseconds
+def binance_timestamp(when):
+    return int((when.timestamp() * 1000) + (when.microsecond / 1000))
+
+
+class Client:
+
+    def __init__(self) -> None:
+        self._sesh = asks.Session(connections=4)
+        self._sesh.base_location = _url
+
+    async def _api(
+        self,
+        method: str,
+        data: dict,
+    ) -> Dict[str, Any]:
+        resp = await self._sesh.get(
+            path=f'/api/v3/{method}',
+            params=data,
+            timeout=float('inf')
+        )
+        return resproc(resp, log)
+
+    async def symbol_info(
+        self,
+        sym: Optional[str] = None
+
+    ) -> dict:
+
+        resp = await self._api('exchangeInfo', {})
+        if sym is not None:
+            for sym_info in resp['symbols']:
+                if sym_info['symbol'] == sym:
+                    return sym_info
+            else:
+                raise SymbolNotFound(f'{sym} not found')
+        else:
+            return resp['symbols']
+
+    async def bars(
+        self,
+        symbol: str = 'BTCUSDT',
+        start_time: int = None,
+        end_time: int = None,
+        limit: int = 1000,  # <- max allowed per query
+        as_np: bool = True,
+
+    ) -> dict:
+
+        if start_time is None:
+            start_time = binance_timestamp(
+                arrow.utcnow().floor('minute').shift(minutes=-limit)
+            )
+
+        if end_time is None:
+            end_time = binance_timestamp(arrow.utcnow())
+
+        # https://binance-docs.github.io/apidocs/spot/en/#kline-candlestick-data
+        bars = await self._api(
+            'klines',
+            {
+                'symbol': symbol,
+                'interval': '1m',
+                'startTime': start_time,
+                'endTime': end_time,
+                'limit': limit
+            }
+        )
+
+        # TODO: pack this bars scheme into a ``pydantic`` validator type:
+        # https://binance-docs.github.io/apidocs/spot/en/#kline-candlestick-data
+
+        # TODO: we should port this to ``pydantic`` to avoid doing
+        # manual validation ourselves..
+        new_bars = []
+        for i, bar in enumerate(bars):
+
+            bar = OHLC(*bar)
+
+            row = []
+            for j, (name, ftype) in enumerate(_ohlc_dtype[1:]):
+
+                # TODO: maybe we should go nanoseconds on all
+                # history time stamps?
+                if name == 'time':
+                    # convert to epoch seconds: float
```

Comment:
We should make an issue about possibly moving to nanoseconds as `int` for all time stamps.
I see no reason to not be thinking about UHFT down the road.